// json-rpc-protocol/dist/format.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.response = exports.request = exports.notification = exports.error = undefined;

var _errors = require('./errors');

// ===================================================================

var toJson = JSON.stringify;

// ===================================================================

// Note: `id` may be null if the identifier of the corresponding
// quests could not be detected.
var error = exports.error = function error(id, _error) {
  // Hide internal errors.
  _error = (typeof _error.toJsonRpcError === 'function' ? _error : new _errors.JsonRpcError()).toJsonRpcError();

  return toJson({
    jsonrpc: '2.0',
    id,
    error: _error
  });
};

// -------------------------------------------------------------------

var notification = exports.notification = function notification(method) {
  var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : undefined;
  return toJson({
    jsonrpc: '2.0',
    method: method,
    params: params
  });
};

// -------------------------------------------------------------------

var request = exports.request = function request(id, method) {
  var params = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : undefined;
  return toJson({
    jsonrpc: '2.0',
    method: method,
    params: params,
    id: id
  });
};

// -------------------------------------------------------------------

var response = exports.response = function response(id, result) {
  return toJson({
    jsonrpc: '2.0',
    id: id,
    result: result
  });
};
